x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java

```diff
@@ -617,44 +622,111 @@ private static void appendCommaSeparatedSet(Set<String> items, StringBuilder sb)
         ClusterState clusterState = clusterService.state();
         Set<String> openJobIds = openJobIds(clusterState);
+
+        logger.info("Updating process for calendar change: {} calendar job IDs, {} open jobs", calendarJobIds.size(), openJobIds.size());
+
         if (openJobIds.isEmpty()) {
             updateListener.onResponse(Boolean.TRUE);
             return;
         }
 
+        // Respond immediately to prevent API timeouts
+        updateListener.onResponse(Boolean.TRUE);
+        // Continue with background processing
+        processCalendarUpdatesAsync(calendarJobIds, openJobIds);
+    }
+
+    private void processCalendarUpdatesAsync(List<String> calendarJobIds, Set<String> openJobIds) {
         boolean appliesToAllJobs = calendarJobIds.stream().anyMatch(Metadata.ALL::equals);
         if (appliesToAllJobs) {
-            submitJobEventUpdate(openJobIds, updateListener);
+            logger.info("Calendar change applies to all jobs - starting background update for {} jobs", openJobIds.size());
+            submitJobEventUpdateAsync(openJobIds);
             return;
         }
-
         // calendarJobIds may be a group or job
-        jobConfigProvider.expandGroupIds(
-            calendarJobIds,
-            updateListener.delegateFailureAndWrap(
-                (delegate, expandedIds) -> threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
-                    // Merge the expanded group members with the request Ids.
-                    // Ids that aren't jobs will be filtered by isJobOpen()
-                    expandedIds.addAll(calendarJobIds);
-                    openJobIds.retainAll(expandedIds);
-                    submitJobEventUpdate(openJobIds, delegate);
-                })
-            )
-        );
+        // Process group expansion asynchronously
+        jobConfigProvider.expandGroupIds(calendarJobIds, ActionListener.wrap(expandedIds -> {
+            expandedIds.addAll(calendarJobIds);
+            openJobIds.retainAll(expandedIds);
+            logger.info("Calendar change expanded to {} jobs - starting background update", openJobIds.size());
+            submitJobEventUpdateAsync(openJobIds);
+        }, e -> logger.error("Failed to expand calendar job groups for background update", e)));
     }
 
-    private void submitJobEventUpdate(Collection<String> jobIds, ActionListener<Boolean> updateListener) {
-        for (String jobId : jobIds) {
-            updateJobProcessNotifier.submitJobUpdate(
-                UpdateParams.scheduledEventsUpdate(jobId),
-                ActionListener.wrap(
-                    isUpdated -> auditor.info(jobId, Messages.getMessage(Messages.JOB_AUDIT_CALENDARS_UPDATED_ON_PROCESS)),
-                    e -> logger.error("[" + jobId + "] failed submitting process update on calendar change", e)
-                )
+    private void submitJobEventUpdateAsync(Collection<String> jobIds) {
```

Comment:
Use a set to ensure the same job is not updated twice

```suggestion
    private void submitJobEventUpdateAsync(Set<String> jobIds) {
```

Reply:
Changed